docker/machine.go

```diff
@@ -0,0 +1,246 @@
+func NewMachine(cluster, machine string, logger logr.Logger) (*Machine, error) {
+	if cluster == "" {
+		return nil, errors.New("cluster is required when creating a docker.Machine")
+	}
+	if machine == "" {
+		return nil, errors.New("machine is required when creating a docker.Machine")
+	}
+	if logger == nil {
+		return nil, errors.New("logger is required when creating a docker.Machine")
+	}
+
+	container, err := getContainer(
+		withLabel(clusterLabel(cluster)),
+		withName(machineContainerName(cluster, machine)),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Machine{
+		cluster:   cluster,
+		machine:   machine,
+		Logger:    logger,
+		container: container,
+	}, nil
+}
+
+// ContainerName return the name of the container for this machine
+func (m *Machine) ContainerName() string {
+	return machineContainerName(m.cluster, m.machine)
+}
+
+// ProviderID return the provider identifier for this machine
+func (m *Machine) ProviderID() string {
+	return fmt.Sprintf("docker:////%s", m.ContainerName())
+}
+
+// NodeName return the name of the Kubernetes node corresponding to this machine
+func (m *Machine) NodeName() string {
+	return m.ContainerName()
+}
+
+func machineContainerName(cluster, machine string) string {
+	return fmt.Sprintf("%s-%s", cluster, machine)
+}
+
+// CreateControlPlane creates a docker container hosting a Kubernetes node.
+func (m *Machine) CreateControlPlane(version *string) error {
+	// Create if not exists.
+	if m.container == nil {
+		var err error
+		m.Info("Creating controlplane machine container")
+		m.container, err = nodes.CreateControlPlaneNode(
+			m.ContainerName(),
+			machineImage(version),
+			clusterLabel(m.cluster),
+			"127.0.0.1",
+			0,
+			nil,
+			nil,
+		)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+// CreateWorker creates a docker container hosting a Kubernetes node.
+func (m *Machine) CreateWorker(version *string) error {
+	// Create if not exists.
+	if m.container == nil {
+		var err error
+		m.Info("Creating worker machine container")
+		m.container, err = nodes.CreateWorkerNode(
+			m.ContainerName(),
+			machineImage(version),
+			clusterLabel(m.cluster),
+			nil,
+			nil,
+		)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+// ExecBootstrap runs bootstrap on a node, this is generally `kubeadm <init|join>`
+func (m *Machine) ExecBootstrap(data string) error {
+	if m.container == nil {
+
+	}
+
+	cloudConfig, err := base64.StdEncoding.DecodeString(data)
+	if err != nil {
+		return errors.Wrap(err, "failed to decode machine's bootstrap data")
+	}
+
+	m.Info("Running machine bootstrap scripts")
+	lines, err := cloudinit.Run(cloudConfig, m.container.Cmder())
+	if err != nil {
+		for _, line := range lines {
+			fmt.Println(line)
+		}
+		return errors.Wrap(err, "failed to join a control plane node with kubeadm")
+	}
+
+	return nil
+}
+
+// SetNodeProviderID patches a node with docker://node-name as the providerID
```

Comment:
this shouldn't reference the direct provider-id format directly. How about `SetNodeProviderID sets the docker provider ID for the kubernetes node`?